### PR TITLE
koord-manager: cleanup node.alloc when device cr have been deleted

### PR DESCRIPTION
**Pull request by lucming**
### Ⅰ. Describe what this PR does
related pr: https://github.com/koordinator-sh/koordinator/pull/1594

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 1]
```diff
@@ -136,6 +136,51 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
 	return err
 }
 
+func (r *NodeResourceReconciler) cleanupGPUNodeResource(node *corev1.Node) error {
+	deletedKeys := []corev1.ResourceName{
+		extension.ResourceGPU,
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/device_resource_calculator.go:141]: Not only the resources of these GPUs, but also RDMA, FPGA, etc.

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 2]
```diff
@@ -136,6 +136,51 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	}
+	needUpdte := false
+	for resourceName, _ := range node.Status.Allocatable {
+		for _, key := range deletedKeys {
+			if key == resourceName {
+				needUpdte = true
+				break
+			}
+		}
+	}
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/device_resource_calculator.go:154]:
```suggestion
	for _, key := range deletedKeys {
	       if _, ok := node.Status.Allocatable[key]; ok {
	           needUpdate = true
	           break
	       }
		}
```

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 3]
```diff
@@ -136,6 +136,51 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	if !needUpdte {
+		return nil
+	}
+
+	err := util.RetryOnConflictOrTooManyRequests(func() error {
+		updateNode := &corev1.Node{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: node.Name}, updateNode); err != nil {
+			klog.Errorf("failed to get node %v, error: %v", node.Name, err)
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/device_resource_calculator.go:162]: recommend use structure log, klog.ErrorS

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ func (r *NodeResourceReconciler) updateDeviceResources(node *corev1.Node) error
 		if !errors.IsNotFound(err) {
 			return fmt.Errorf("failed to get device resources: %w", err)
 		}
-		return nil
+		return r.cleanupGPUNodeResource(node)
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/device_resource_calculator.go:44]: we should add UT for the change.

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/slo-controller/noderesource/device_resource_calculator.go (hunk 1 above)
Comment (lucming) [reply pkg/slo-controller/noderesource/device_resource_calculator.go:143]: RDMA, FPGA have been added.
RDMA, FPGA are not reported by koordlet now,right? other plugin?

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 5]
```diff
@@ -136,6 +136,51 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
 }
 
+func (r *NodeResourceReconciler) cleanupGPUNodeResource(node *corev1.Node) error {
+	deletedKeys := []corev1.ResourceName{
+		extension.ResourceGPU,
+		extension.ResourceGPUCore,
+		extension.ResourceGPUMemory,
+		extension.ResourceGPUMemoryRatio,
+	}
+	needUpdte := false
```
[the remark this answers is not in the document]
Comment (lucming) [reply pkg/slo-controller/noderesource/device_resource_calculator.go]: done

File: pkg/slo-controller/noderesource/device_resource_calculator.go (hunk 2 above)
Comment (lucming) [reply pkg/slo-controller/noderesource/device_resource_calculator.go]: done

File: pkg/slo-controller/noderesource/device_resource_calculator.go (hunk 4 above)
Comment (lucming) [reply pkg/slo-controller/noderesource/device_resource_calculator.go]: done

File: pkg/slo-controller/noderesource/device_resource_calculator.go (hunk 1 above)
Comment (eahydra) [reply pkg/slo-controller/noderesource/device_resource_calculator.go:143]: yes，koordlet does not report these devices but other plugins will report them。

Comment (lucming) [reply pkg/slo-controller/noderesource/device_resource_calculator.go:143]: I've wondered before which resources need to be cleaned up, device crd has extended interfaces for other types of devices, code in https://github.com/koordinator-sh/koordinator/blob/main/apis/scheduling/v1alpha1/device_types.go#L26-L30 ,  but koord-manager only handles gpu type devices, code in https://github.com/koordinator-sh/koordinator/blob/main/pkg/slo-controller/noderesource/device_resource_calculator.go#L37-L59 
If there are plans to handle other devices via koordinater, they can also be left in, or I'll comment them out first, WDYT?

Comment (eahydra) [reply pkg/slo-controller/noderesource/device_resource_calculator.go:143]: Just deal with the GPU first.

**Comment by lucming** [comment]
Done.-------- 原始邮件 --------发件人： Joseph ***@***.***>日期： 2023年9月1日周五 11:58收件人： koordinator-sh/koordinator ***@***.***>抄送： lucming ***@***.***>, Author ***@***.***>主    题： Re: [koordinator-sh/koordinator] koord-manager: cleanup node.alloc when device cr have been deleted (PR #1597)
@eahydra commented on this pull request.



In pkg/slo-controller/noderesource/device_resource_calculator.go:
> @@ -136,6 +136,51 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
 	return err
 }
 
+func (r *NodeResourceReconciler) cleanupGPUNodeResource(node *corev1.Node) error {
+	deletedKeys := []corev1.ResourceName{
+		extension.ResourceGPU,

Just deal with the GPU first.

**Comment by eahydra** [comment]
/lgtm

PTAL @zwzhang0107 @saintube

**Review by eahydra — APPROVED** [review]

**Review by jasonliu747 — APPROVED** [review]

**Pull request merged**